src/Microsoft.AspNetCore.Routing/Internal/UriBuildingContext.cs

```diff
@@ -27,20 +29,30 @@ public class UriBuildingContext
         public UriBuildingContext(UrlEncoder urlEncoder)
         {
             _urlEncoder = urlEncoder;
-            _uri = new StringBuilder();
+            _path = new StringBuilder();
+            _query = new StringBuilder();
             _buffer = new List<BufferValue>();
-            Writer = new StringWriter(_uri);
+            PathWriter = new StringWriter(_path);
+            QueryWriter = new StringWriter(_query);
```

Comment:
Are the path and query written out of order? Is it not possible to get away with one builder & writer and keep an index of when the path stops and the query begins?

Reply:
That's true, but do need to ToString them separately though, and then combine the result.

Reply:
That's intended though (at least for now) so that we can use `UriHelper` instead of creating our own thing.